### PR TITLE
Feature/multiline

### PR DESCRIPTION
**Pull request by kingwingfly**
Hi! Your code is really in high quality!

I added a new feature `multiline` to this crate. 

Now it looks like: 

Screenshot 2024-05-30 at 07 52 17

I have done following in this PR:
1. multiline support
2. an example `cargo run --example multiline_input --features multiline`
3. navigate between lines with arrow-up and down (maybe not well enough)
4. enable `doc_auto_cfg` for better document with features
5. move doc in `lib.rs` to `src/DOC.md` and include it with `include_str!`

How:
1. Add two flag to `CursorString`: `multiline` and `editing`. (only when `multiline` feature enabled)
2. If not in editing mode, insert nothing. (Initially `editing` is true)
3. When Enter hit, try insert `\n`
4. If `cursor.multiline` is false, do not insert `\n`
5. Only `cursor.multiline` is true, `Tab` can switch `editing` flag between true and false. This ensure one-line input works as past.

### CODE REVIEW

**Comment by fadeevab** [comment]
@kingwingfly Hi! Thanks, awesome :) I will try to review ASAP according to my availability.

File: build.rs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+use rustc_version::{version_meta, Channel};
+
+fn main() {
+    // Set cfg flags depending on release channel
+    if let Channel::Nightly = version_meta().unwrap().channel {
+        println!("cargo::rustc-check-cfg=cfg(CHANNEL_NIGHTLY)");
+        println!("cargo:rustc-cfg=CHANNEL_NIGHTLY");
+    }
```
Comment (fadeevab) [inline build.rs:8]: It must be a matter of a separate PR.

**Review by fadeevab — COMMENTED** [review]
Other than that:
1. Make the hints dimmed.
2. Is it possible to implement Shift+Enter for submitting to avoid the "view mode"?

File: build.rs (hunk 1 above)
Comment (kingwingfly) [reply build.rs]: I tried `console` and even `crossterm` by just printing the key I pressed, they can recognize many keys modified by `Shift`, but `Enter`. 

The strange thing is, `Ctrl` `Alt` `Shift` work for many keys, but no matter which modifier I'm pressing, those crates just recognize `Enter` as long as I press it.

I'll take a look at `zed`

**Comment by kingwingfly** [comment]
I just found, on Windows powershell, `console` treats `Enter` as `Key::Enter` and `Ctrl+Enter` as `Key::Unknown + Key::Char('\n')`.
On macOS, it's always `Key::Enter`...

**Comment by kingwingfly** [comment]
I've tried my best. [This](https://users.rust-lang.org/t/how-to-capture-ctrl-enter-keyboard-event-in-console-on-macos/112292/6) contains the conclusion.  

- It's nearly impossible to got the state of Ctrl from terminal. 
- It can be implemented with `winapi` on Windows,  `x11` on Linux, `cocoa` on macOS,
- but it's not worth trying for a CLI tool, and it's useless with remote ssh.

**Comment by fadeevab** [comment]
@kingwingfly What about to consider `Alt+Enter`?
**Linux**: `Key::UnknownEscSeq(['\r'])`
**Windows**: `Key::Alt` (for whatever reason)
**MacOs**: ?

File: src/DOC.md [hunk 2]
```diff
@@ -0,0 +1,259 @@
+Beautiful, minimal, opinionated CLI prompts inspired by the
```
Comment (fadeevab) [inline src/DOC.md:1]: I mean, need to revert this refactoring in this PR fully, it's another topic to consider, but it's not about multiline input

**Comment by fadeevab** [comment]
@kingwingfly Oh, for Windows, when it returns `Alt`, you read the second key and it's `Enter`
image

**Comment by kingwingfly** [comment]
```rust
use console::{Key, Term};

fn main() {
    let term = Term::stdout();

    while let Ok(key) = term.read_key() {
        dbg!(&key);
        if key == Key::Char('q') {
            break;
        }
        if key == Key::Alt {
            println!("Alt");    // unreachable on macOS
        }
    }
}
```

| platform | key | result |
| -- | -- | -- |
| macOS| Enter | Enter |
| macOS| AnyModifier + Enter | Enter |
| Windows | Enter | Enter |
| Windows | Ctrl + Enter | Unknown + Char('\n') |
| Windows | Alt + Enter | Alt + Char('\n') |
| macOS ssh to Windows | - | Same as on macOS | 
| macOS use Remote Desktop to Windows | _ | Same as on Windows but `Alt + Enter` is FullScreen shortcut on macOS  |

For mac users, it seems only `Enter` keyboard-event can be read from console no matter what modifier is.

File: src/prompt/cursor.rs [hunk 3]
```diff
@@ -2,10 +2,27 @@ use std::fmt::{Display, Formatter, Result};
 
 use zeroize::ZeroizeOnDrop;
 
-#[derive(Default, ZeroizeOnDrop, Clone)]
+#[derive(ZeroizeOnDrop, Clone)]
+#[cfg_attr(not(feature = "multiline"), derive(Default))]
 pub struct StringCursor {
     value: Vec<char>,
     cursor: usize,
+    #[cfg(feature = "multiline")]
```
Comment (fadeevab) [inline src/prompt/cursor.rs:10]: @kingwingfly  Currently I'm thinking about how could your patch be improved with more decoupling from the main source code and simplification.
First thing about the cursor: I believe `editing` is not the property of the cursor, it's property of the input prompt (UI). And `multiline` could be "always enabled" cause it's about the `\n` in the string: it is inserted from the outside or not. Removing these two fields (moving them outside of the cursor) will make it simpler and context-independent. Also, less of cfg switches.

Comment (kingwingfly) [reply src/prompt/cursor.rs:10]: I tried your suggestion.

First, I remove the two flags from `StringCursor`, and add `editing` to `Input`. This leads that I cannot switch `editing` flag correctly and decide whether to render footer without `multiline` flag.  And `Tab` will always switch editing flag on one-line input. 

Then I add `multiline` flag to `Input`. But `PromptInteraction::interact_on_prepared` cannot got `self.multiline` because it's in trait. It will always insert `\n`. 

Then I move this job to `PromptInteraction::on`. It's used to decide what's the next State, but I modify input there, that's strange. 

Finally, I think `multiline` and `editing` should be the attributes of a `StringCursor`. `multiline` stands for whether it accept '\n'. `editing` stands for whether it's mutable. 

I'd like to make these two flags public instead...

Comment (kingwingfly) [reply src/prompt/cursor.rs:10]: > multiline could be "always enabled"

I misunderstood this before. (I used to think this asks to remove flag in StringCursor)

I thought keeping `multiline` as feature may be better since it add two byte to Input, and many may not want multiline at all. 

I'm not a CS student and don't have any practical experience. This may be quite performance anxiety, if you insist to make it always enable and remove all `cfg feature`, I'll try tonight.

Comment (fadeevab) [reply src/prompt/cursor.rs:10]: @kingwingfly your code doesn't degrade performance, it only contributes to the binary size... so, we might even consider to remove the feature flag. Anyways, I need to take time to play with your patch to not load you with pointless tasks.

Comment (fadeevab) [reply src/prompt/cursor.rs:10]: @kingwingfly if you have capacity,  could you check how much the binary size of a basic example increases with the multiline feature enabled and how longer is the build time?

Comment (kingwingfly) [reply src/prompt/cursor.rs:10]: Okey, I believe there's little influence on performance or binary size. 

But here are two facts as so far:
1. If feature removed, always enable, for those who don't need `multiline`, there would be two useless flags, `multiline` and `editing`.
2. The two flag can be in `Input` or `StringCursor`, and both ways work. However, the frontier needs modify `Input.input` in `PromptInteraction::on`. Although other prompts like `Select` changes cursor in `on`, I prefer gathering input-modifying in `PromptInteraction::interact_on_prepared` as before, and `on` only for state transfer.

So, there are two question left:
1. Whether enable `multiline` by default and remove the feature flag or not?
2. `multiline` and `editing` flags should belong to `Input` or `StringCursor`?

**Comment by fadeevab** [comment]
@kingwingfly I tested your patch, pay attention to the visual effects while manipulating the cursor (don't pay attention to the trailing line, it's the bug of the `console` + my terminal):

cliclack-multiline

**Comment by kingwingfly** [comment]
Reproduce method:
- add an empty line at last
- move cursor up
- the last empty line is missed

I try fix then.

**Comment by kingwingfly** [comment]
I fixed last line missing bug.

After that, I mainly fixed bugs with interactively validation, and add more tests to `example/multiline_input.rs`. 

However, I found this looks not so pretty:
image

I'd like the key hint to be fixed at botton-right:
image


But this needs `Term::size`, with `reder()` refactored. And it may also leads width usize overflow bugs, complicated.

**Comment by fadeevab** [comment]
@kingwingfly I'd rather keep only the yellow part, as it means that the next move is to edit the field (thus, switching to the edit mode automatically would be good).

BTW, I tried to refactor your PR locally in the way I suggested: and it looks fine for me, I want to cherry-pick some of your fresh patches and post it. Also, I found that removing the feature flag makes the patch simpler and less prone to errors like missing the validation handling.

**Comment by fadeevab** [comment]
@kingwingfly This is how I believe it should look like: https://github.com/fadeevab/cliclack/pull/54/files

It doesn't contain all bugfixes though.

**Comment by kingwingfly** [comment]
@fadeevab Nice refactor!!!

I just reviewed and feel it can work good. I'll test it in 24 hours.

**Comment by kingwingfly** [comment]
We are working on this instead

 - #54
